chore: add analytics documentation (#515)
front-commerce/developers.front-commerce.com: docs/advanced/analytics/plugins.mdx

---

> sidebar_position: 2

Reviewer: **nitpick:**  I think that the tracking API is more important than plugins.\
Hypothesis: an integrator or newcomer might more often want to see how to track new events than creating a new plugin

**suggestion:** swap sidebar positions

title: Plugins
description:
  The analytics library is designed to work with any third party analytics tools
  via plugins.
---

import AnalyticsFigure from "./assets/analytics.svg";
import Figure from "@site/src/components/Figure";
import ContactLink from "@site/src/components/ContactLink";
import { useState } from "react";

> _Since version 2.19_

<p>{frontMatter.description}</p>

Plugins are a powerful abstraction that let you:

> - add a new analytics provider (like google analytics or pixel)

Reviewer: **question:** is "pixel" the name of "Meta pixel"? I think it isn't (but really not sure) and it  seems weird in this sentence.

**suggestion:** for consistency, rename with the exact service name
```suggestion
- add a new analytics provider (like Google analytics or Meta pixel)
```

- hook into an existing analytics provider plugin
- or add any kind of logic to react to visitor actions

Plugins can be broken down into 2 types:

- **Provider plugins** - connecting to third party analytic services
- **Custom plugins** - additional features, data manipulation, & any other side
  effects.

Both have the same signature, as illustrated below.

> 

Reviewer: **nitpick:** it isn't super clear that these plugins are _just_  analytics plugins

**suggestion:** link to https://getanalytics.io/plugins so users can find the canonical documentation for analytics plugins and maybe also to https://getanalytics.io/plugins/writing-plugins/ from the same block (or in the section below?)

1. Have a loaded function to let analytics know when its safe to send the third
   party data.

Here is an example if the `analytics-plugin-acme` plugin:

```js title="my-analytics-module/plugins/analytics-plugin-acme.js"
// `settings` is the value defined in the plugin config in `config/analytics.js`
export default function providerPluginExample(settings) {
  // return object for analytics to use
  return {
    /* All plugins require a name */
    name: "acme",
    /* Everything else below this is optional depending on your plugin requirements */
    config: {
      whatEver: settings.whatEver,
      elseYouNeed: settings.elseYouNeed,
    },
    initialize: ({ config }) => {
      // load provider script to page
    },
    page: ({ payload }) => {
      // call provider specific page tracking
    },
    track: ({ payload }) => {
      // call provider specific event tracking
    },
    identify: ({ payload }) => {
      // call provider specific user identify method
    },
    loaded: () => {
      // return boolean so analytics knows when it can send data to third party
      return !!window.myPluginLoaded;
    },
  };
}
```

### Extending with `EcommercePlugin`

> The `EcommercePlugin` plugin will allow you to create an interface to map

Reviewer: ```suggestion
The `EcommercePlugin` plugin is a plugin provided by Front-Commerce. It allows you to create an interface to map
```

[e-commerce events and properties](https://developers.google.com/analytics/devguides/collection/ga4/ecommerce)
to the relevant events and properties for your `acme` provider, you can find a
full list of events in the
[EcommercePlugin.js](https://gitlab.blackswift.cloud/front-commerce/front-commerce/-/blob/main/src/web/core/analytics/plugins/e-commerce/domain/e-commerce-plugin.js)

Below is an example of how to extend the `analytics-plugin-acme` plugin to

> handle e-commerce events.

Reviewer: **issue:** I've struggled to understand in how it relates to `analytics-plugin-acme` from reading the example.\
Couldn't it be splitted into 2 different sections:
1. write the plugin
2. use this plugin as an enhancer of an existing one

(could also work in reverse order BTW)

        // the tracking method will be called with these properties
        properties: {
           product_id: properties.sku,
           review_rating: properties.rating,
        },
      }
    },
    "Product Shared": (properties) => {
        return {
          event:"acme_product_shared",
          properties: {
             product_id: properties.sku,
             method: properties.method,
          },
        }
    },
  };
}

export default function providerPluginExample(settings) {
  // wrap your config with the ecommerce plugin
  return new AcmeEcommercePlugin({
    // ... same config as defined in the previous example
  };)
}
```

<div>

:::info

We handle this internally for
[known plugins](https://gitlab.blackswift.cloud/front-commerce/front-commerce/-/blob/main/src/web/core/analytics/plugins/e-commerce/e-commerce.js#L6-11)
like the [`google-analytics`](https://getanalytics.io/plugins/google-analytics/)
plugin.

If you are aware of any other providers which require this,
please&nbsp;<ContactLink>let us know</ContactLink> or feel free to contribute to
our
[analytics examples](https://gitlab.blackswift.cloud/front-commerce/examples/analytics).

:::

</div>

> ## Using Plugins

Reviewer: **suggestion:** maybe moving this section after the intro could help readers to better understand what `settings` are in the section that explains how to write your  own one.

loading for your end users.

Let's add our `analytics-plugin-acme` plugin to our `config/analytics.js` file.

```js title="src/config/analytics.js"
module.exports = {
  analytics: {
    // ...config
    plugins: [
      // ...plugins
     {
        // remember to add the `acme` to your cookieServices config or disable the needConsent
        name: "acme",
        needConsent: true,
        // all these settings will be passed to your plugin script
        settings: {
          whatEver: "foo",
          elseYouNeed: "bar",
        }
        // add a dynamic import to load the plugin script
        script: () => import("my-analytics-module/plugins/analytics-plugin-acme.js"),
      },
    ],
  },
};
```

:::info REMINDER

Just like any other integrations, don't forget to setup the `cookiesServices.js`
file accordingly, in order to load the newly created integration only when the
user has given their consent.

:::

### Custom plugins common pain points

In some cases, the documentation for implementing the tracking might not be
compatible with <abbr title="Single Page Application">SPA</abbr>'s
architectures. This can be the case if they are asking you to add a `script` tag

> in your head, or at the end of your `body` without any further information.

Reviewer 2: ```suggestion
in your `head`, or at the end of your `body` without any further information.
```

    userConsents: otherAuthorizations
  }
  script: () => import("@segment/analytics.js-integration-google-tag-manager"),
}
```

Update your <abbr title="Content Security Policy">CSP</abbr> in
`src/config/website.js` according to the tags you use (see
[Google Tag Manager's documentation](https://developers.google.com/tag-platform/tag-manager/web/csp)
for more details)

In GTM, you will then be able to leverage several specific things configured in
your plugins.

First, the `userConsents` configuration option will be pushed to your dataLayer
as the `userConsents` value. You can reference it from a Variable in GTM. Here
is an example:

<Figure>

![Screenshot of a GTM Variable configured to expose the user consents](./assets/gtm-datalayer-variable.png)

</Figure>

Then, you can leverage the `UserConsentUpdated` event tracked whenever users
update their consent preferences. You could create triggers to enable scripts to
load / remove (depending on the `userConsents` value). Here is an example:

<Figure>

![Screenshot of a GTM Trigger configured to detect when users gave their consent to a specific integration](./assets/gtm-trigger-example.png)

> </Figure>

Author: I need to verify this behavior as I have not yet understood what/how the datalayer worked before, so requires some investigation, 
which is attached to a window object and retrieved via the [dataLayerName](https://getanalytics.io/plugins/google-tag-manager/#configuration-options-for-browser) config key for the analytics.io plugin

> </Figure>

Author: I need to verify this behavior as I have not yet understood what/how the datalayer worked before, so requires some investigation, 

which is attached to a window object and retrieved via the [dataLayerName](https://getanalytics.io/plugins/google-tag-manager/#configuration-options-for-browser) config key for the analytics.io plugin

> ### Pixel

Reviewer: **issue:** As per my previous comment, I think we should be specific to the pixel used

```suggestion
### Meta Pixel
```

…  by the way, it seems that the official name is `Meta` all over the place :roll_eyes:\
The repository should be renamed too don't you agree?
